:100644 100644 fec591b... 0000000... M	Makefile.am.inc

diff --git a/Makefile.am.inc b/Makefile.am.inc
index fec591b..9f14c29 100644
--- a/Makefile.am.inc
+++ b/Makefile.am.inc
@@ -1,2 +1,2 @@
-AM_CFLAGS = -Wall -Werror -Wno-error=deprecated-declarations -Wextra -Wwrite-strings -Wno-unused-parameter -std=gnu99 -D_GNU_SOURCE -D_REENTRANT
+AM_CFLAGS = -Wall -Werror -Wno-error=deprecated-declarations -Wextra -Wwrite-strings -Wno-unused-but-set-variable -Wno-unused-parameter -std=gnu99 -D_GNU_SOURCE -D_REENTRANT
 
